Uninstall clears NO_PROXY env variable, which it should not

On our site we have some RSFA based K2S testagents to use. These have the the following proxy settings set
  Setx /m HTTP_PROXY http://pxbud.evosoft.com:8080 
  Setx /m HTTPS_PROXY http://pxbud.evosoft.com:8080 
  Setx /m NO_PROXY 172.19.1.0/24,172.20.0.0/16,172.21.0.0/16,192.168.137.*,172.19.1.*,172.20.1.*,172.21.1.*,172.22.1.*,carbonk8s.dev,127.0.0.1,%COMPUTERNAME%,local,svc.cluster.local,HUBUDO4BUILD01,HUBUDO4BUILD02,EVH03286WS,EVH03279WS

Since the update to 1.7.1, we have observed, that the "k2s install" does not run due to the following error message.

  ##[error]Exception caught by K2s cluster and addons installer: Error or warning when installing K2s.
  ≡ƒñû Installing K2s 'k2s' v1.7.1 in 'C:\k2s\1.7.1\core' on Microsoft Windows 11 Iot Enterprise Ltsc 2024 10.0.26100.4061 Build 26100.4061
  ΓÅ│ [18:03:43] Directory 'C:\ProgramData\K2s' not existing, creating it ...
  ΓÅ│ [18:03:43] Directory 'C:\ProgramData\K2s' created.
  ΓÅ│ [ProxyValidation] Invalid proxy environment variable configuration:
  ΓÅ│ [Machine] NO_PROXY is missing required K2s hosts: .cluster.local, .local, ::1, 127.0.0.1, 172.19.1.100, 172.20.0.0/16, 
  ΓÅ│ 172.21.0.0/16, localhost. These hosts must be excluded from the proxy to allow K2s cluster communication.
  ΓÅ│ At C:\k2s\1.7.1\core\lib\modules\k2s\k2s.node.module\windowsnode\proxy\proxy.module.psm1:458 char:9
  ΓÅ│ +         throw "[ProxyValidation] Invalid proxy environment variable c ...
  ΓÅ│ +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ΓÅ│     + CategoryInfo          : OperationStopped: ([ProxyValidatio... communication.:String) [], RuntimeException
  ΓÅ│     + FullyQualifiedErrorId : [ProxyValidation] Invalid proxy environment variable configuration:
  ΓÅ│     [Machine] NO_PROXY is missing required K2s hosts: .cluster.local, .local, ::1, 127.0.0.1, 172.19.1.100, 172.20.0.0 
  ΓÅ│    /16, 172.21.0.0/16, localhost. These hosts must be excluded from the proxy to allow K2s cluster communication.

 Even if we added the required entries to the machine level environment variable, so it looked like this
  Setx /m NO_PROXY .cluster.local,.local,::1,127.0.0.1,172.19.1.100,172.20.0.0/16,172.21.0.0/16,localhost,172.19.1.0/24,172.20.0.0/16,172.21.0.0/16,192.168.137.*,172.19.1.*,172.20.1.*,172.21.1.*,172.22.1.*,carbonk8s.dev,127.0.0.1,%COMPUTERNAME%,local,svc.cluster.local,HUBUDO4BUILD01,HUBUDO4BUILD02,EVH03286WS,EVH03279WS
we got successful testrun, but for the second testrun the variable was cleaned up, and second run failed again with the error message above. See https://apollo.siemens-healthineers.com/tfs/IKM.TPC.Projects/syngoONE/_build/results?buildId=158328…

We have reset the environment variable by hand, and have observed the following modifications done to the no_proxy variable


1. test execute pipeline - passed (modified value seems to be the same as the current value)
https://apollo.siemens-healthineers.com/tfs/IKM.TPC.Projects/syngoONE/_build/results?buildId=158335…
pipeline start: 2026-05-28T15:25:30.4061741Z

5/28/2026 3:27:12.0452977 PM
powershell "&'C:\k2s\1.7.1\core\lib\scripts\k2s\base\Invoke-ExecScript.ps1' -Script \"&'C:\k2s\1.7.1\core\lib\scripts\k2s\install\install.ps1' -MasterVMProcessorCount 4 -MasterVMMemory 24GB -MasterDiskSize 50GB -AdditionalHooksDir '$Env:MED_CONFIG\Deploy\Scripts\Hooks\RSFA' -ShowLogs -SkipStart\""
Type:  REG_SZ
Length:  536
Data:  .cluster.local,.local,::1,127.0.0.1,172.19.1.*,172.19.1.0/24,172.19.1.100,172.20.0.0/16,172.20.1.*,172.21.0.0/16,172.21.1.*,172.22.1.*,192.168.137.*,carbonk8s.dev,EVH03279WS,EVH03286WS,HUBUDO4BUILD01,HUBUDO4BUILD02,local,localhost,O4VMK2STA02,svc.cluster.local

pipeline end: 2026-05-28T18:00:41.0584259Z

2. test pipeline - failed with no_proxy error
https://apollo.siemens-healthineers.com/tfs/IKM.TPC.Projects/syngoONE/_build/results?buildId=158335…
pipeline start: 2026-05-28T18:01:07.2396710Z

5/28/2026 6:02:35.4577899 PM
powershell "&'C:\k2s\1.7.1\core\lib\scripts\k2s\base\Invoke-ExecScript.ps1' -Script \"&'C:\k2s\1.7.1\core\lib\scripts\k2s\uninstall\uninstall.ps1'\""
Type:  REG_SZ
Length:  362
Data:  172.19.1.*,172.19.1.0/24,172.20.1.*,172.21.1.*,172.22.1.*,192.168.137.*,carbonk8s.dev,EVH03279WS,EVH03286WS,HUBUDO4BUILD01,HUBUDO4BUILD02,local,O4VMK2STA02,svc.cluster.local

5/28/2026 6:03:26.8767365 PM
powershell "&'C:\k2s\1.7.1\core\lib\scripts\k2s\base\Invoke-ExecScript.ps1' -Script \"&'C:\k2s\1.7.1\core\lib\scripts\k2s\uninstall\uninstall.ps1'\""
Type:  REG_SZ
Length:  362
Data:  172.19.1.*,172.19.1.0/24,172.20.1.*,172.21.1.*,172.22.1.*,192.168.137.*,carbonk8s.dev,EVH03279WS,EVH03286WS,HUBUDO4BUILD01,HUBUDO4BUILD02,local,O4VMK2STA02,svc.cluster.local

5/28/2026 6:04:20.9403162 PM
powershell "&'C:\k2s\1.7.1\core\lib\scripts\k2s\base\Invoke-ExecScript.ps1' -Script \"&'C:\k2s\1.7.1\core\lib\scripts\k2s\uninstall\uninstall.ps1'\""
Type:  REG_SZ
Length:  362
Data:  172.19.1.*,172.19.1.0/24,172.20.1.*,172.21.1.*,172.22.1.*,192.168.137.*,carbonk8s.dev,EVH03279WS,EVH03286WS,HUBUDO4BUILD01,HUBUDO4BUILD02,local,O4VMK2STA02,svc.cluster.local
pipeline end: 2026-05-28T18:04:42.1272527Z

To us, it seems that the uninstall script is cleaning up entries from the NO_PROXY environment variable, that the install needs.

--- a/lib/modules/k2s/k2s.node.module/windowsnode/proxy/proxy.module.psm1
+++ b/lib/modules/k2s/k2s.node.module/windowsnode/proxy/proxy.module.psm1
@@ -9,6 +9,7 @@ $configModule = "$PSScriptRoot\..\..\..\k2s.infra.module\config\config.module.ps
 Import-Module $configModule
 
 $proxyConfigFp = "C:\ProgramData\k2s\proxy.conf"
+$noProxyBaselineFp = "C:\ProgramData\k2s\no_proxy.baseline"
 
 class ProxyConfig {
     [string] $HttpProxy
@@ -56,6 +57,57 @@ function Get-K2sHosts {
     $k2sHosts = $k2sHosts | Where-Object { ![string]::IsNullOrWhiteSpace($_) } | Sort-Object -Unique
     
     return $k2sHosts
+}
+
+function ConvertTo-NoProxyList {
+    param (
+        [AllowNull()]
+        [object[]]$Entries
+    )
+
+    if ($null -eq $Entries -or $Entries.Count -eq 0) {
+        return @()
+    }
+
+    return $Entries |
+        ForEach-Object { $_ } |
+        ForEach-Object { $_ -split ',' } |
+        ForEach-Object { $_.Trim() } |
+        Where-Object { ![string]::IsNullOrWhiteSpace($_) } |
+        Sort-Object -Unique
+}
+
+function Save-NoProxyBaseline {
+    param (
+        [AllowNull()]
+        [string]$NoProxyValue
+    )
+
+    if (Test-Path -Path $noProxyBaselineFp) {
+        return
+    }
+
+    $baselineDirectory = Split-Path -Path $noProxyBaselineFp -Parent
+    if (!(Test-Path -Path $baselineDirectory)) {
+        New-Item -ItemType Directory -Path $baselineDirectory | Out-Null
+    }
+
+    $baseline = ConvertTo-NoProxyList -Entries @($NoProxyValue)
+    $baseline -join ',' | Set-Content -Path $noProxyBaselineFp
+}
+
+function Get-NoProxyBaseline {
+    if (!(Test-Path -Path $noProxyBaselineFp)) {
+        return $null
+    }
+
+    return ConvertTo-NoProxyList -Entries @((Get-Content -Path $noProxyBaselineFp -Raw))
+}
+
+function Remove-NoProxyBaseline {
+    if (Test-Path -Path $noProxyBaselineFp) {
+        Remove-Item -Path $noProxyBaselineFp -Force
+    }
 }
 
 <#
@@ -501,10 +553,9 @@ function Add-K2sHostsToNoProxyEnvVar() {
     $k2sHosts = Get-K2sHosts
 
     if (![string]::IsNullOrWhiteSpace($noProxyEnvVar)) {
-        $noProxyList = $noProxyEnvVar -split ","
+        Save-NoProxyBaseline -NoProxyValue $noProxyEnvVar
 
-        $noProxyList += $k2sHosts
-        $noProxyList = $noProxyList | Sort-Object -Unique
+        $noProxyList = ConvertTo-NoProxyList -Entries @($noProxyEnvVar, $k2sHosts)
         $updatedNoProxyEnvVar = $noProxyList -join ","
         [Environment]::SetEnvironmentVariable("NO_PROXY", $updatedNoProxyEnvVar, "Process")
         [Environment]::SetEnvironmentVariable("NO_PROXY", $updatedNoProxyEnvVar, "Machine")
@@ -523,12 +574,21 @@ function Remove-K2sHostsFromNoProxyEnvVar() {
     $k2sHosts = Get-K2sHosts
 
     if (![string]::IsNullOrWhiteSpace($noProxyEnvVar)) {
-        $noProxyList = $noProxyEnvVar -split ","
-        $noProxyList = $noProxyList | Where-Object { $_ -notin $k2sHosts }
+        $noProxyBaseline = Get-NoProxyBaseline
+        if ($null -eq $noProxyBaseline) {
+            Write-Log '[Proxy] NO_PROXY baseline not found. Keeping machine NO_PROXY unchanged.' -Console
+            return
+        }
+
+        $noProxyList = ConvertTo-NoProxyList -Entries @($noProxyEnvVar)
+        $k2sAddedHosts = $k2sHosts | Where-Object { $_ -notin $noProxyBaseline }
+        $noProxyList = $noProxyList | Where-Object { $_ -notin $k2sAddedHosts }
         $updatedNoProxyEnvVar = $noProxyList -join ","
         [Environment]::SetEnvironmentVariable("NO_PROXY", $updatedNoProxyEnvVar, "Process")
         [Environment]::SetEnvironmentVariable("NO_PROXY", $updatedNoProxyEnvVar, "Machine")
     }
+
+    Remove-NoProxyBaseline
 }
 
 Export-ModuleMember -Function Get-OrUpdateProxyServer,

--- a/lib/modules/k2s/k2s.node.module/windowsnode/proxy/proxy.module.unit.tests.ps1
+++ b/lib/modules/k2s/k2s.node.module/windowsnode/proxy/proxy.module.unit.tests.ps1
@@ -558,6 +558,9 @@ Describe 'Add-K2sHostsToNoProxyEnvVar' -Tag 'unit', 'ci', 'proxy' {
     It 'adds K2s hosts to existing NO_PROXY environment variable' {
         InModuleScope $moduleName {
             Mock Get-K2sHosts { return @('localhost', '127.0.0.1', '.local') }
+            $script:noProxyBaselineFp = Join-Path ([System.IO.Path]::GetTempPath()) "k2s-no-proxy-baseline-add-$PID.txt"
+            $originalMachineNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+            $originalProcessNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Process")
             
             [Environment]::SetEnvironmentVariable("NO_PROXY", "existing.com,another.com", "Machine")
             
@@ -569,8 +572,34 @@ Describe 'Add-K2sHostsToNoProxyEnvVar' -Tag 'unit', 'ci', 'proxy' {
                 $result | Should -Match 'existing.com'
             }
             finally {
-                [Environment]::SetEnvironmentVariable("NO_PROXY", $null, "Machine")
-                [Environment]::SetEnvironmentVariable("NO_PROXY", $null, "Process")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalMachineNoProxy, "Machine")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalProcessNoProxy, "Process")
+                Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
+            }
+        }
+    }
+
+    It 'stores the original NO_PROXY value before adding K2s hosts' {
+        InModuleScope $moduleName {
+            Mock Get-K2sHosts { return @('localhost', '127.0.0.1') }
+            $script:noProxyBaselineFp = Join-Path ([System.IO.Path]::GetTempPath()) "k2s-no-proxy-baseline-store-$PID.txt"
+            $originalMachineNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+            $originalProcessNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Process")
+            
+            [Environment]::SetEnvironmentVariable("NO_PROXY", " existing.com, localhost ", "Machine")
+            
+            try {
+                Add-K2sHostsToNoProxyEnvVar
+                $baseline = Get-Content -Path $script:noProxyBaselineFp -Raw
+                $result = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+                
+                $baseline.Trim() | Should -Be 'existing.com,localhost'
+                $result | Should -Be '127.0.0.1,existing.com,localhost'
+            }
+            finally {
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalMachineNoProxy, "Machine")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalProcessNoProxy, "Process")
+                Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
             }
         }
     }
@@ -578,21 +607,84 @@ Describe 'Add-K2sHostsToNoProxyEnvVar' -Tag 'unit', 'ci', 'proxy' {
 }
 
 Describe 'Remove-K2sHostsFromNoProxyEnvVar' -Tag 'unit', 'ci', 'proxy' {
-    It 'removes K2s hosts from NO_PROXY environment variable' {
+    It 'preserves K2s hosts that existed before K2s changed NO_PROXY' {
         InModuleScope $moduleName {
             Mock Get-K2sHosts { return @('localhost', '127.0.0.1', '.local') }
+            $script:noProxyBaselineFp = Join-Path ([System.IO.Path]::GetTempPath()) "k2s-no-proxy-baseline-preserve-$PID.txt"
+            $originalMachineNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+            $originalProcessNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Process")
             
+            Set-Content -Path $script:noProxyBaselineFp -Value 'localhost,existing.com'
             [Environment]::SetEnvironmentVariable("NO_PROXY", "localhost,existing.com,127.0.0.1,.local", "Machine")
             
             try {
                 Remove-K2sHostsFromNoProxyEnvVar
                 
                 $result = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
-                $result | Should -Be 'existing.com'
+                $result | Should -Be 'existing.com,localhost'
+                Test-Path -Path $script:noProxyBaselineFp | Should -Be $false
             }
             finally {
-                [Environment]::SetEnvironmentVariable("NO_PROXY", $null, "Machine")
-                [Environment]::SetEnvironmentVariable("NO_PROXY", $null, "Process")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalMachineNoProxy, "Machine")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalProcessNoProxy, "Process")
+                Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
+            }
+        }
+    }
+
+    It 'keeps required K2s hosts across an add and remove lifecycle when they were pre-existing' {
+        InModuleScope $moduleName {
+            Mock Get-K2sHosts { return @('localhost', '127.0.0.1', '::1', '172.19.1.100', '172.20.0.0/16', '172.21.0.0/16', '.local', '.cluster.local') }
+            $script:noProxyBaselineFp = Join-Path ([System.IO.Path]::GetTempPath()) "k2s-no-proxy-baseline-lifecycle-$PID.txt"
+            $originalMachineNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+            $originalProcessNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Process")
+            $preExistingNoProxy = '.cluster.local,.local,::1,127.0.0.1,172.19.1.100,172.20.0.0/16,172.21.0.0/16,localhost,172.19.1.0/24,carbonk8s.dev,svc.cluster.local'
+            
+            [Environment]::SetEnvironmentVariable("NO_PROXY", $preExistingNoProxy, "Machine")
+            
+            try {
+                Add-K2sHostsToNoProxyEnvVar
+                Remove-K2sHostsFromNoProxyEnvVar
+                
+                $result = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine") -split ','
+                $result | Should -Contain '.cluster.local'
+                $result | Should -Contain '.local'
+                $result | Should -Contain '::1'
+                $result | Should -Contain '127.0.0.1'
+                $result | Should -Contain '172.19.1.100'
+                $result | Should -Contain '172.20.0.0/16'
+                $result | Should -Contain '172.21.0.0/16'
+                $result | Should -Contain 'localhost'
+                $result | Should -Contain 'carbonk8s.dev'
+                $result | Should -Contain 'svc.cluster.local'
+            }
+            finally {
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalMachineNoProxy, "Machine")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalProcessNoProxy, "Process")
+                Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
+            }
+        }
+    }
+
+    It 'keeps NO_PROXY unchanged when no baseline is available' {
+        InModuleScope $moduleName {
+            Mock Get-K2sHosts { return @('localhost', '127.0.0.1', '.local') }
+            $script:noProxyBaselineFp = Join-Path ([System.IO.Path]::GetTempPath()) "k2s-no-proxy-baseline-missing-$PID.txt"
+            $originalMachineNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine")
+            $originalProcessNoProxy = [Environment]::GetEnvironmentVariable("NO_PROXY", "Process")
+            
+            Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
+            [Environment]::SetEnvironmentVariable("NO_PROXY", "localhost,existing.com,127.0.0.1,.local", "Machine")
+            
+            try {
+                Remove-K2sHostsFromNoProxyEnvVar
+                
+                [Environment]::GetEnvironmentVariable("NO_PROXY", "Machine") | Should -Be 'localhost,existing.com,127.0.0.1,.local'
+            }
+            finally {
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalMachineNoProxy, "Machine")
+                [Environment]::SetEnvironmentVariable("NO_PROXY", $originalProcessNoProxy, "Process")
+                Remove-Item -Path $script:noProxyBaselineFp -Force -ErrorAction SilentlyContinue
             }
         }
     }

